lib/cache/health_check_config.go

```diff
@@ -0,0 +1,112 @@
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package cache
+
+import (
+	"context"
+
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport/api/defaults"
+	headerv1 "github.com/gravitational/teleport/api/gen/proto/go/teleport/header/v1"
+	healthcheckconfigv1 "github.com/gravitational/teleport/api/gen/proto/go/teleport/healthcheckconfig/v1"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/api/utils"
+	"github.com/gravitational/teleport/api/utils/clientutils"
+	"github.com/gravitational/teleport/lib/services"
+)
+
+const healthCheckConfigStoreNameIndex = "name"
+
+func newHealthCheckConfigCollection(upstream services.HealthCheckConfigReader, w types.WatchKind) (*collection[*healthcheckconfigv1.HealthCheckConfig], error) {
+	if upstream == nil {
+		return nil, trace.BadParameter("missing parameter HealthCheckConfigReader")
+	}
+
+	return &collection[*healthcheckconfigv1.HealthCheckConfig]{
+		store: newStore(map[string]func(*healthcheckconfigv1.HealthCheckConfig) string{
+			healthCheckConfigStoreNameIndex: func(r *healthcheckconfigv1.HealthCheckConfig) string {
+				return r.GetMetadata().GetName()
+			},
+		}),
+		fetcher: func(ctx context.Context, loadSecrets bool) ([]*healthcheckconfigv1.HealthCheckConfig, error) {
+			var out []*healthcheckconfigv1.HealthCheckConfig
+			clientutils.IterateResources(ctx,
+				upstream.ListHealthCheckConfigs,
+				func(hcc *healthcheckconfigv1.HealthCheckConfig) error {
+					out = append(out, hcc)
+					return nil
+				},
+			)
+			return out, nil
+		},
+		headerTransform: func(hdr *types.ResourceHeader) *healthcheckconfigv1.HealthCheckConfig {
+			return &healthcheckconfigv1.HealthCheckConfig{
+				Kind:    hdr.Kind,
+				Version: hdr.Version,
+				Metadata: &headerv1.Metadata{
+					Name: hdr.Metadata.Name,
+				},
+			}
+		},
+		watch: w,
+	}, nil
+}
+
+// ListHealthCheckConfigs lists health check configs with pagination.
+func (c *Cache) ListHealthCheckConfigs(ctx context.Context, pageSize int, nextToken string) ([]*healthcheckconfigv1.HealthCheckConfig, string, error) {
+	ctx, span := c.Tracer.Start(ctx, "cache/ListHealthCheckConfigs")
+	defer span.End()
+
+	out, next, err := genericList(ctx,
+		pageSize,
+		nextToken,
+		genericLister[*healthcheckconfigv1.HealthCheckConfig]{
+			cache:           c,
+			collection:      c.collections.healthCheckConfig,
+			index:           healthCheckConfigStoreNameIndex,
+			defaultPageSize: defaults.DefaultChunkSize,
+			upstreamList:    c.Config.HealthCheckConfig.ListHealthCheckConfigs,
+			nextToken: func(t *healthcheckconfigv1.HealthCheckConfig) string {
+				return t.GetMetadata().GetName()
+			},
+			clone: func(s *healthcheckconfigv1.HealthCheckConfig) *healthcheckconfigv1.HealthCheckConfig {
+				return utils.CloneProtoMsg(s)
+			},
```

Comment:
Does this compile?
```suggestion
			clone: utils.CloneProtoMsg,
```

Reply:
Better yet, if we know that these types are actual protobuf types:
```suggestion
			clone: proto.CloneOf,
```